src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts

```diff
@@ -412,32 +472,59 @@ export class OsdListComponent implements OnInit {
     });
   }
 
+  /**
+   * Perform check first and display a critical confirmation modal.
+   * @param {string} actionDescription name of the action.
+   * @param {string} itemDescription the item's name that the action operates on.
+   * @param {string} templateItemDescription the action name to be displayed in modal template.
+   * @param {Function} check the function is called to check if the action is safe.
+   * @param {string} checkKey the safe indicator's key in the check response.
+   * @param {Function} action the action function.
+   * @param {boolean} oneshot if true, action function is called with all items as parameter.
+   *   Otherwise, multiple action functions with individual items are sent.
+   */
   showCriticalConfirmationModal(
     actionDescription: string,
     itemDescription: string,
     templateItemDescription: string,
-    action: (id: number) => Observable<any>
+    check: (ids: number[]) => Observable<any>,
+    checkKey: string,
+    action: (id: number | number[]) => Observable<any>,
+    taskWrapped: boolean = false
   ): void {
-    this.osdService.safeToDestroy(JSON.stringify(this.getSelectedOsdIds())).subscribe((result) => {
+    check(this.getSelectedOsdIds()).subscribe((result) => {
       const modalRef = this.modalService.show(CriticalConfirmationModalComponent, {
         initialState: {
           actionDescription: actionDescription,
           itemDescription: itemDescription,
           bodyTemplate: this.criticalConfirmationTpl,
           bodyContext: {
-            result: result,
+            safeToPerform: result[checkKey],
+            message: result.message,
             actionDescription: templateItemDescription
           },
           submitAction: () => {
-            observableForkJoin(
+            const observable = observableForkJoin(
               this.getSelectedOsdIds().map((osd: any) => action.call(this.osdService, osd))
-            ).subscribe(
-              () => {
-                this.getOsdList();
-                modalRef.hide();
-              },
-              () => modalRef.hide()
             );
+            if (taskWrapped) {
+              observable.subscribe(
+                undefined,
+                () => {
+                  this.getOsdList();
+                  modalRef.hide();
+                },
+                () => modalRef.hide()
+              );
+            } else {
+              observable.subscribe(
+                () => {
+                  this.getOsdList();
+                  modalRef.hide();
+                },
+                () => modalRef.hide()
+              );
+            }
```

Comment:
Can't this be merged into a single observable and then subscribed just once? I just grepped our code and see ~449 subscriptions and only 40 un-subscriptions :worried:  That means a lot of memory leaks. I'd suggest not to add 3 new ones every time someone clicks on the delete button. You may push subscriptions to a `private subscriptions: Subscription[]` and on ngDestroy deallocate those `this.forEach(s => s.unsubscribe())`.

Reply:
If most of them are finite http observables, they don't need to be manually unsubscribed.

> I just grepped our code and see ~449 subscriptions and only 40 un-subscriptions  That means a lot of memory leaks.

Reply:
Thanks @bk201, didn't know that detail! However we use lots of periodic (infinite) Observables for polling HTTP endpoints, and then we chain other observables to the output of those ones, so I'm still concerned about that.